sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc

```diff
@@ -0,0 +1,321 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#include "opentelemetry/sdk/metrics/aggregation/base2_exponential_histogram_aggregation.h"
+#include <stddef.h>
+#include <stdint.h>
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include "opentelemetry/common/spin_lock_mutex.h"
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/sdk/metrics/aggregation/aggregation.h"
+#include "opentelemetry/sdk/metrics/aggregation/aggregation_config.h"
+#include "opentelemetry/sdk/metrics/aggregation/base2_exponential_histogram_indexer.h"
+#include "opentelemetry/sdk/metrics/data/circular_buffer.h"
+#include "opentelemetry/sdk/metrics/data/metric_data.h"
+#include "opentelemetry/sdk/metrics/data/point_data.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+namespace
+{
+
+uint32_t GetScaleReduction(int32_t start_index, int32_t end_index, size_t max_buckets) noexcept
+{
+  uint32_t scale_reduction = 0;
+  while (static_cast<size_t>(end_index - start_index + 1) > max_buckets)
+  {
+    start_index >>= 1;
+    end_index >>= 1;
+    scale_reduction++;
+  }
+  return scale_reduction;
+}
+
+// Commented out as it is not used
+// uint32_t GetScaleReduction(const AdaptingCircularBufferCounter &first,
+//                            const AdaptingCircularBufferCounter &second,
+//                            size_t max_buckets)
+// {
+//   if (first.Empty() || second.Empty())
+//   {
+//     return 0;
+//   }
+//   const int32_t start_index = (std::min)(first.StartIndex(), second.StartIndex());
+//   const int32_t end_index   = (std::max)(first.EndIndex(), second.EndIndex());
+//   return GetScaleReduction(start_index, end_index, max_buckets);
+// }
+
+void DownscaleBuckets(AdaptingCircularBufferCounter *buckets, uint32_t by) noexcept
+{
+  if (buckets->Empty())
+  {
+    return;
+  }
+
+  // We want to preserve other optimisations here as well, e.g. integer size.
+  // Instead of  creating a new counter, we copy the existing one (for bucket size
+  // optimisations), and clear the values before writing the new ones.
+  // TODO(euroelessar): Do downscaling in-place.
+  AdaptingCircularBufferCounter new_buckets = *buckets;
```

Comment:
yes, this would be costly operation, specifically when the actual aggregation happens under mutex lock (though that needs to change too), and downscaling is not rare. Do we have any benchmark on throughput when multiple threads are trying to aggregate. Not a blocker for this PR, but better to create a separate issue to track this.

Reply:
By benchmark, I was thinking of something along the lines of what’s done in [measurements_benchmark.cc](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/sdk/test/metrics/measurements_benchmark.cc). Would be good to capture this in a separate issue for follow-up.

Reply:
Thank you for the pointer. @felipesantosk and I will address in-place downscaling + multi-threaded benchmarking in a new issue + follow-up PR.

Reply:
@lalitb created new issue to track this #3366